File: server/services/multilingual_agent.py
"""
Multilingual Agent - Provides content translation across supported languages
"""
import json
from typing import Dict, List, Any
from shared_config import (
    model, is_valid_api_key, logger, TranslationRequest, 
    parse_gemini_response, genai_types
)

async def translate_content(request: TranslationRequest) -> Dict[str, Any]:
    """Translate content using Gemini API"""
    if not model:
        return {
            "originalText": request.text,
            "translatedText": f"[TRANSLATED FROM {request.fromLanguage.upper()} TO {request.toLanguage.upper()}] {request.text}",
            "fromLanguage": request.fromLanguage,
            "toLanguage": request.toLanguage,
            "note": "This is demo content. Configure a valid Gemini API key for real translations."
        }
    
    try:
        prompt = f"""Translate the following text from {request.fromLanguage} to {request.toLanguage}. 
        
Provide accurate, culturally appropriate translation suitable for Indian educational context.

**Requirements:**
1. Maintain educational meaning and context
2. Use appropriate cultural references for Indian students
3. Preserve technical terms where appropriate
4. Include pronunciation guide if helpful
5. Note any cultural adaptations made

Text to translate: "{request.text}"

**Output Format (JSON):**
{{
  "originalText": "original text",
  "translatedText": "accurately translated text",
  "fromLanguage": "source language",
  "toLanguage": "target language",
  "culturalNotes": "any cultural adaptation notes if relevant",
  "pronunciationGuide": "pronunciation help if needed",
  "alternativeTranslations": ["alternative ways to express the same meaning"],
  "contextualUsage": "when and how to use this translation in educational settings",
  "difficulty": "language difficulty level for students"
}}

Ensure the translation is natural and appropriate for Indian educational context!"""

        response = model.generate_content(
            prompt,
            generation_config=genai_types.GenerationConfig(
                temperature=0.3,
                max_output_tokens=1024,
            )
        )
        
        result = parse_gemini_response(response.text)
        
        # Add metadata
        result["aiGenerated"] = True
        result["model"] = "gemini-1.5-pro-latest"
        
        logger.info(f"Successfully translated from {request.fromLanguage} to {request.toLanguage}")
        return result
        
    except Exception as error:
        logger.error(f"Gemini API Error: {error}")
        return {
            "originalText": request.text,
            "translatedText": f"[TRANSLATION ERROR] {request.text}",
            "fromLanguage": request.fromLanguage,
            "toLanguage": request.toLanguage,
            "note": f"Translation failed due to API error: {error}"
        }

File: server/services/multilingual-agent.ts
import { model, isValidApiKey, logger, TranslationRequest, parseGeminiResponse } from './shared-config';

// Supported languages mapping
const SUPPORTED_LANGUAGES = {
  'english': 'en',
  'hindi': 'hi',
  'tamil': 'ta',
  'telugu': 'te',
  'bengali': 'bn',
  'marathi': 'mr',
  'gujarati': 'gu',
  'kannada': 'kn',
  'malayalam': 'ml',
  'punjabi': 'pa',
  'urdu': 'ur',
  'sanskrit': 'sa'
};

function generateMockTranslation(request: TranslationRequest): any {
  /**Generate mock translation when Gemini is not available*/
  return {
    originalText: request.text,
    translatedText: `[TRANSLATED FROM ${request.fromLanguage.toUpperCase()} TO ${request.toLanguage.toUpperCase()}] ${request.text}`,
    fromLanguage: request.fromLanguage,
    toLanguage: request.toLanguage,
    confidence: 0.8,
    note: "Demo translation - Gemini service unavailable"
  };
}

function validateLanguages(fromLanguage: string, toLanguage: string): boolean {
  const fromLang = fromLanguage.toLowerCase();
  const toLang = toLanguage.toLowerCase();
  
  return Object.keys(SUPPORTED_LANGUAGES).includes(fromLang) && 
         Object.keys(SUPPORTED_LANGUAGES).includes(toLang);
}

export async function translateContent(request: TranslationRequest): Promise<any> {
  /**Translate content using Multilingual Agent*/
  try {
    if (!validateLanguages(request.fromLanguage, request.toLanguage)) {
      logger.error(`Unsupported language pair: ${request.fromLanguage} to ${request.toLanguage}`);
      return {
        error: "Unsupported language pair",
        supportedLanguages: Object.keys(SUPPORTED_LANGUAGES)
      };
    }

    if (!isValidApiKey || !model) {
      logger.warn("Gemini API not configured, using mock content");
      return generateMockTranslation(request);
    }

    const prompt = `Translate the following text from ${request.fromLanguage} to ${request.toLanguage}.
    
    Text to translate: "${request.text}"
    
    Requirements:
    - Maintain the original meaning and context
    - Use appropriate cultural context for Indian languages
    - Preserve any educational terminology
    - Ensure the translation is natural and fluent
    
    Return the response as a JSON object with the following structure:
    {
      "originalText": "Original text",
      "translatedText": "Translated text",
      "fromLanguage": "Source language",
      "toLanguage": "Target language",
      "confidence": 0.95,
      "culturalNotes": "Any cultural context notes",
      "alternatives": ["alternative translation 1", "alternative translation 2"]
    }`;

    const result = await model.generateContent(prompt);
    const response = await result.response;
    const text = response.text();
    
    const parsedResponse = parseGeminiResponse(text);
    
    if (parsedResponse.error) {
      logger.error(`Failed to parse Gemini response: ${parsedResponse.error}`);
      return generateMockTranslation(request);
    }
    
    return {
      ...parsedResponse,
      fromLanguage: request.fromLanguage,
      toLanguage: request.toLanguage
    };
    
  } catch (error) {
    logger.error(`Multilingual Agent Error: ${error}`);
    return generateMockTranslation(request);
  }
}

export function getSupportedLanguages(): string[] {
  return Object.keys(SUPPORTED_LANGUAGES);
} 